docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,363 @@
+    // AddComponentVersion adds a new component version to the repository.
+    // If a component version already exists, it will be updated with the new descriptor.
+    // The descriptor internally will be serialized via the runtime package.
+    // The descriptor MUST have its target Name and Version already set as they are used to identify the target
+    // Location in the Store.
+    AddComponentVersion(ctx context.Context, descriptor *descriptor.Descriptor) error
+}
+
+
+type TargetRepositoryProvider interface {
+    // GetTargetRepository returns the target ocm component version repository 
+    // for the given component specification in the constructor.
+    GetTargetRepository(ctx context.Context, *constructor.Component) (TargetRepository, error)
+}
+
+type ResourceRepository interface {
+    // DownloadResource downloads a resource from the repository.
+    DownloadResource(ctx context.Context, res *descriptor.Resource) (content blob.ReadOnlyBlob, err error)
+}
+
+type ResourceRepositoryProvider interface {
+    // ResourceRepositoryProvider returns the target ocm resource repository for the given component specification in the constructor.
+    GetResourceRepository(ctx context.Context, *constructor.Component) (Repository, error)
+}
+
+// Options are the options for construction based on a *constructor.Constructor.
+type Options struct {
+    // While constructing a component version, the constructor library will use the given target repository provider
+    // to get the target repository for the component specification.
+    TargetRepositoryProvider
+    
+    // While constructing a component version, the constructor library will use the given resource repository provider
+    // to get the resource repository for the component specification when processing resources by value.
+    ResourceRepositoryProvider
+    
+    // While constructing a component version, the constructor library will use the given resource input method provider
+    // to get the resource input method for the component specification when processing resources with an input method.
+    ResourceInputMethodProvider
+    
+    // While constructing a component version, the constructor library will use the given source input method provider
+    // to get the source input method for the component specification when processing sources with an input method.
+    SourceInputMethodProvider
+
+    // While constructing a component version, the constructor library will use the 
+    // given function to decide whether a resource should be processed by value or not.
+    ProcessByValue func(*constructor.Resource) bool
+}
+
+// Construct constructs a component version based on the given specification and options.
+func Construct(ctx context.Context, specification *constructor.Constructor, opts Options) ([]*descriptor.Descriptor, error)
```

Comment:
What is `*constructor.Constructor`? Did I miss the definition of that?

Reply:
The Constructor is part of the specification as part of https://github.com/open-component-model/open-component-model/pull/111. The package is quite large since it is mostly a redefinition of the component descriptors so thats why I didnt add it to the contract. Do you prefer me to add that here?

Reply:
Hmm, hmmm. Maybe just something small, you know? I'm thinking that reading an ADR should be possible to understand things without the need to navigate away? Not sure.

Reply:
I added a reference to the specification as a separate section as it was too huge for the ADR to include the entire spec